### PR TITLE
esp32s2: Add initial support for ESP32-S2

### PR DESCRIPTION
**Pull request by mrrosen**
Add initial support for ESP32-S2, another low-cost WiFi SoC from Espressif (https://www.espressif.com/en/products/socs/esp32-s2). Requires update to include ESP IDF (https://github.com/zephyrproject-rtos/esp-idf/pull/3).

Initial board: https://docs.espressif.com/projects/esp-idf/en/latest/esp32s2/hw-reference/esp32s2/user-guide-saola-1-v1.2.html

### CODE REVIEW

File: boards/xtensa/esp32s2/Kconfig.board
Comment (dcpleung) [inline boards/xtensa/esp32s2/Kconfig.board:7]: Maybe change `ESP32S2` to `ESP32-S2` for the prompt so it is consistent with the documentation below.

**Review by dcpleung — CHANGES_REQUESTED** [review]
Looks like tabs was replaced with spaces in various places. Could you change them back to tabs?

File: boards/xtensa/esp32s2/Kconfig.board
Comment (mrrosen) [reply boards/xtensa/esp32s2/Kconfig.board:7]: Sure, I tried doing that everywhere user-visible and missed this one

**Comment by mrrosen** [comment]
Ive fixed the files Ive found where my editor used spaces instead of tabs; and Ive installed the pre-commit style check hook so hopefully that wont happen again :)

**Review by galak — COMMENTED** [review]
Is it possible to look at adding devicetree support for the interrupt controller (need on esp32 as well)?

File: west.yml
Comment (galak) [inline west.yml:47]: You probably want this to be `pull/3/head`

https://docs.zephyrproject.org/latest/guides/modules.html?highlight=modules#process

File: boards/xtensa/esp32s2/esp32s2_defconfig
Comment (jli157) [inline boards/xtensa/esp32s2/esp32s2_defconfig:8]: What special requirements are needed to specify the main stack size in the board defconfig file?

File: west.yml
Comment (mrrosen) [reply west.yml]: Fixed

File: boards/xtensa/esp32s2/esp32s2_defconfig
Comment (mrrosen) [reply boards/xtensa/esp32s2/esp32s2_defconfig:8]: Copied from ESP32's board def; so it comes from there. Seems like alot of boards do this even though its probably best if they didnt; but that might be a good change in another PR

**Comment by mrrosen** [comment]
I think Ive fixed all the issues the CI except a few - new file needs an owner and some generated files are missing licences (these dont have licenses for esp32 either, so what should be done?)

**Comment by dcpleung** [comment]
`_soc_inthandlers.h` is a generated file so it could be alright without license. However, could you fix the issue in the "Compliance" check? You can add yourself to the CODEOWNERS file for the dtsi file.

**Comment by mrrosen** [comment]
@dcpleung Im noticing now that the original esp32 doesnt have a code owner for its dts; is there actually an owner for this file that would be better to take over ownership for both?

**Comment by dcpleung** [comment]
> @dcpleung Im noticing now that the original esp32 doesnt have a code owner for its dts; is there actually an owner for this file that would be better to take over ownership for both?

Since you add the new files, please make yourself the code owner for now. We can deal with the leftover later.

**Review by dcpleung — APPROVED** [review]

**Comment by andrewboie** [comment]
needs a rebase, and the compliance / scan code issues addressed.

**Comment by mrrosen** [comment]
@andrewboie I wasnt pushing on this as it requires an update to the ESP IDF included with Zephyr to 4.2, which add support for ESP32-S2 but has officially been released yet, so I was waiting for Espressif to formally release 4.2 before updating this PR and the one on the Zephyr ESP IDF to make sure they were based on the official 4.2 release.

As for the scan issues, for the license scan, it is only failing on generated files which dont have licenses for ESP32 (should we add licenses to these files?) and for the compliance scan it seems to only be looking at the first versions of the commits which I amended to fix the Signed-off-by in the current version

**Comment by andrewboie** [comment]
OK, let's wait until 4.2 arrives.

File: soc/xtensa/esp32s2/linker.ld
Comment (tejlmand) [inline soc/xtensa/esp32s2/linker.ld:1]: This file is identical i layout to this file:
https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/xtensa/esp32/linker.ld

But with different values on some lines.

Could you align, so that a single file can be used ?
Similar to what is done for:
ARC: https://github.com/zephyrproject-rtos/zephyr/blob/master/include/arch/arc/v2/linker.ld
ARM: https://github.com/zephyrproject-rtos/zephyr/blob/master/include/arch/arm/aarch32/cortex_m/scripts/linker.ld
and others.

and then include, for example like here:
https://github.com/zephyrproject-rtos/zephyr/blob/8d563d65074c26e6a19cace753fe9558fd28f323/soc/arc/snps_arc_hsdk/linker.ld#L23
and
https://github.com/zephyrproject-rtos/zephyr/blob/8d563d65074c26e6a19cace753fe9558fd28f323/soc/arm/atmel_sam/sam3x/linker.ld#L9

**Review by tejlmand — CHANGES_REQUESTED** [review]
Two main points are the linker script, and the ExternalProject usage.

File: cmake/toolchain/espressif/generic.cmake
Comment (mrrosen) [reply cmake/toolchain/espressif/generic.cmake]: Fixed in latest push

File: boards/xtensa/esp32s2/CMakeLists.txt
Comment (mrrosen) [reply boards/xtensa/esp32s2/CMakeLists.txt]: From copying esp32's board file, fixed in latest push

File: boards/xtensa/esp32s2/CMakeLists.txt
Comment (mrrosen) [reply boards/xtensa/esp32s2/CMakeLists.txt:19]: So, this is taken directly from esp32 board CMakeLists.txt for building their own bootloader. Im sure there is a better way of doing it now that I look at it more closely but Ill work to fix that once esp idf v4.2 comes out.

File: soc/xtensa/esp32s2/linker.ld
Comment (mrrosen) [reply soc/xtensa/esp32s2/linker.ld:1]: So, all Xtensa targets right now do have very similar linking scripts (by nature of being the same architecture), not just esp32 and esp32s2, and they do the same thing (for example, setup the vector table) in slightly different ways. I can look into unifying them in the same way ARC and ARM are; but it will require a bit more research into the possible configuration options since Xtensa is a pretty dynamic architecture dependent on the configuration settings when generating the IP.

Comment (mrrosen) [reply soc/xtensa/esp32s2/linker.ld:1]: Should that be a separate PR on just the existing Xtensa architectures Zephyr supports or should that be part of this one?

File: drivers/gpio/gpio_esp32.c
Comment (ExtremeGTX) [inline drivers/gpio/gpio_esp32.c:67]: Since these values are the same as ESP32, it should be moved out `#elif` section

File: boards/xtensa/esp32s2/esp32s2.dts
Comment (ExtremeGTX) [inline boards/xtensa/esp32s2/esp32s2.dts:27]: Why you didn't use PLL settings, is the clock_driver not compatible with ESP32-S2?

**Review by ExtremeGTX — APPROVED** [review]

File: soc/xtensa/esp32s2/linker.ld
Comment (tejlmand) [reply soc/xtensa/esp32s2/linker.ld:1]: For aligning esp32 and esp32s2 then doing so in this PR fine.
A more general cleanup should be in independent cleanup.

File: boards/xtensa/esp32s2/CMakeLists.txt
Comment (tejlmand) [reply boards/xtensa/esp32s2/CMakeLists.txt:19]: it seems someone else is working on this:
https://github.com/zephyrproject-rtos/zephyr/pull/30424

**Comment by sylvioalves** [comment]
@mrrosen do you have plans to work with S2 support?

**Comment by mrrosen** [comment]
> @mrrosen do you have plans to work with S2 support?

At present, no, I had more free time when I did this initial port but have since been busy with other things.

**Pull request closed without merging**
